x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -398,4 +406,52 @@ private FileChannelReference acquireFileChannelReference() {
         ensureOpen();
         return tracker.getAbsentRangeWithin(start, end);
     }
+
+    // used in tests
+    boolean isFSynced() {
+        return fsynced.get();
+    }
+
+    /**
+     * Ensure that all ranges of data written to the cache file are written to the storage device that contains it. This method returns the
+     * list of all successfully written ranges of data since the creation of the cache file.
+     *
+     * @return the list of ranges of data available in cache at the time this method is invoked
+     * @throws IOException                       if the cache file failed to be fsync
+     * @throws AlreadyClosedException            if the cache file is evicted
+     * @throws java.nio.file.NoSuchFileException if the cache file does not exist
+     */
+    public List<Tuple<Long, Long>> fsync() throws IOException {
+        ensureOpen();
+        if (refCounter.tryIncRef()) {
+            try {
+                // Capture the completed ranges before fsyncing; ranges that are completed after this point won't be considered as
+                // persisted on disk by the caller of this method, even if they are fully written to disk at the time the file
+                // fsync is effectively executed
+                final List<Tuple<Long, Long>> completedRanges = tracker.getCompletedRanges();
+                assert completedRanges != null;
+
+                if (fsynced.compareAndSet(false, true)) {
+                    boolean success = false;
+                    try {
+                        // check again if the file is evicted before doing expensive I/O
+                        ensureOpen();
+                        IOUtils.fsync(file, false); // TODO don't forget to fsync parent directory
```

Comment:
Yes, I also thought about borrowing the current file channel if we have one but I found that it added more complexity (mostly in incref/decref the file channel another time) to save a file descriptor that would be quickly released. 

I can maybe give it another try by implementing something that use the file channel if there's one or use IOUtils.fsync() otherwise. Deferring the fsync at closing time does not seem the right thing to do as we don't want to fsync everytime a chunk is written but instead control when cache files are fsynced.